network/ocsp.go

```diff
@@ -0,0 +1,344 @@
+			log.Debugf("OCSP: appending server %s, from cert", serverToCheck.url)
+			serversToCheck = append(serversToCheck, serverToCheck)
+		}
+	} else {
+		if len(cert.OCSPServer) > 0 {
+			log.Debugf("OCSP: Ignoring %d OCSP servers from certificate", len(cert.OCSPServer))
+		}
+	}
+
+	if v.Config.url != "" {
+		serverToCheck := ocspServerToCheck{url: v.Config.url, fromCert: false}
+
+		if v.Config.fromCert == ocspFromCertPrefer || v.Config.fromCert == ocspFromCertTrust {
+			log.Debugf("OCSP: appending server %s, from config", serverToCheck.url)
+			serversToCheck = append(serversToCheck, serverToCheck)
+		} else {
+			log.Debugf("OCSP: prepending server %s, from config", serverToCheck.url)
+			serversToCheck = append([]ocspServerToCheck{serverToCheck}, serversToCheck...)
+		}
+	}
+
+	queriedOCSPs := make(map[string]struct{})
+
+	confirms := 0
+
+	for _, serverToCheck := range serversToCheck {
+		log.Debugf("OCSP: Trying server %s", serverToCheck.url)
+
+		if _, ok := queriedOCSPs[serverToCheck.url]; ok {
+			log.Debugln("OCSP: Skipping, already queried")
+			continue
+		}
+
+		response, err := v.Client.Query(cert.Issuer.CommonName, cert, issuer, serverToCheck.url)
+		if err != nil {
+			log.WithError(err).WithField("url", serverToCheck.url).Warnln("Cannot query OCSP server")
+
+			if v.Config.required == ocspRequiredAll {
+				return ErrOCSPRequiredAllButGotError
+			}
+
+			continue
+		}
+
+		switch response.Status {
+		case ocsp.Good:
+			confirms++
+
+			if serverToCheck.fromCert {
+				log.Debugln("OCSP: confirmed by server from certificate")
+			} else {
+				log.Debugln("OCSP: confirmed by server from config")
+			}
+
+			if v.Config.required != ocspRequiredAll {
+				// One confirmation is enough if we don't require all OCSP servers to confirm the certificate validity
+				break
+			}
+
+			if serverToCheck.fromCert && v.Config.fromCert == ocspFromCertTrust {
+				// If this OCSP server came from certificate and `--tls_ocsp_from_cert=trust`, don't perform further checks
+				break
+			}
+		case ocsp.Revoked:
+			// If any OCSP server replies with "certificate was revoked", return error immediately
+			log.WithField("serial", cert.SerialNumber).WithField("revoked_at", response.RevokedAt).Warnln("OCSP: Certificate was revoked")
+			return ErrCertWasRevoked
+		case ocsp.Unknown:
+			// Treat "Unknown" response as error if tls_ocsp_required is "yes" or "all"
+			if v.Config.required != ocspRequiredNo {
+				log.WithField("url", serverToCheck.url).WithField("serial", cert.SerialNumber).Warnln("OCSP server doesn't know about certificate")
+				return ErrOCSPUnknownCertificate
+			}
+		}
+
+		queriedOCSPs[serverToCheck.url] = struct{}{}
+	}
+
+	if len(serversToCheck) > 0 && confirms == 0 {
+		return ErrOCSPNoConfirms
+	}
+	return nil
+}
+
+// Verify ensures certificate is not revoked by querying configured OCSP servers
+func (v DefaultOCSPVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
```

Comment:
Similarly to CRL, I do not see how `rawCerts` is used in this function. May be we can remote it from input parameters?

Reply:
Yes, with current implementation we don't use `rawCerts`. I left it for possible future use. Here (as well as in `CertVerifier` interface) just to match signature of `VerifyPeerCertificate` callback used in TLS config to call our verifiers.
Do you think we better remove it unless it is really used?

Reply:
@storojs72 , previously @iamnotacake removed it. but we decided to leave it to be compatible with tls standard library in golang. plus it allows using some custom extensions in tls certificates which will not be parsed and used by the standard library. probably it is why golang's tls library pass parsed and raw certificates - https://golang.org/pkg/crypto/tls/#Config